lucene/core/src/java/org/apache/lucene/index/IndexWriterConfig.java

```diff
@@ -463,16 +465,65 @@ public IndexWriterConfig setCommitOnClose(boolean commitOnClose) {
    * Set the {@link Sort} order to use for all (flushed and merged) segments.
    */
   public IndexWriterConfig setIndexSort(Sort sort) {
-    for (SortField sortField : sort.getSort()) {
+    for (SortOrder sortField : sort.getSort()) {
       if (sortField.getIndexSorter() == null) {
         throw new IllegalArgumentException("Cannot sort index with sort field " + sortField);
       }
     }
     this.indexSort = sort;
-    this.indexSortFields = Arrays.stream(sort.getSort()).map(SortField::getField).collect(Collectors.toSet());
+    this.indexSortFields = extractFields(sort);
     return this;
   }
 
+  private Set<String> extractFields(Sort sort) {
+    Set<String> fields = new HashSet<>();
+    for (SortOrder sortOrder : sort.getSort()) {
+      IndexSorter sorter = sortOrder.getIndexSorter();
+      assert sorter != null;
+      try {
+        sorter.getDocComparator(new DocValuesLeafReader() {
```

Comment:
Hmm this is a little bit scary hackity, compared to what we had before :)

Are you wanting to not add a `SortOrder.getField()`?

Reply:
Right, because some sorts refer to multiple fields, or to none.  For example, if we want to add an IndexSorter to expression sorts, we need to check that none of the DV fields referred to in the expression are updateable.

Reply:
Wow, that makes sense -- could you add a comment explaining that the mapping from sort to field is not necessarily "one to one" and so it makes sense we must do something like this to "discover" which fields will be invoked by the sort?  It's still tricky/messy ... I wish we had a cleaner way to ask the `SortOrder` which indexed (doc values?) fields it depends on.